Xamarin.Forms.Platform.Android/ImageCache.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Xamarin.Forms.Platform.Android
+{
+	/// <summary>
+	/// I setup the access to all the cache elements to be async because
+	/// if I didn't then it was locking up the GC and freezing the entire app
+	/// </summary>
+	class ImageCache
+	{
+		readonly FormsLruCache _lruCache;
+		readonly ConcurrentDictionary<string, SemaphoreSlim> _waiting;
+
+		public ImageCache() : base()
+		{
+			_waiting = new ConcurrentDictionary<string, SemaphoreSlim>();
+			_lruCache = new FormsLruCache();			
+		}
+
+		void Put(string key, TimeSpan cacheValidity, global::Android.Graphics.Bitmap cacheObject)
+		{
+			_lruCache.Put(key, new CacheEntry() { TimeToLive = DateTimeOffset.Now.Add(cacheValidity), Data = cacheObject });
+		}
+
+		public Task<Java.Lang.Object> GetAsync(string cacheKey, TimeSpan cacheValidity, Func<Task<Java.Lang.Object>> createMethod)
+		{
+			return Task.Run(async () =>
+			{
+				SemaphoreSlim semaphoreSlim = null;
+
+				try
+				{
+					semaphoreSlim = _waiting.GetOrAdd(cacheKey, (key) => new SemaphoreSlim(1, 1));
+					await semaphoreSlim.WaitAsync().ConfigureAwait(false);
+
+					var cacheEntry = _lruCache.Get(cacheKey) as CacheEntry;
+
+					if (cacheEntry?.TimeToLive < DateTimeOffset.Now)
```

Comment:
Same here? `UtcNow`?

Reply:
I don't think it matters since it's a datetimeoffset and not a datetime it'll do an accurate comparison but I'll set it to utc anyway

Reply:
Changed